imap/mboxname.c

```diff
@@ -2292,11 +2292,32 @@ EXPORTED char *mboxname_conf_getpath(const mbname_t *mbname, const char *suffix)
     char *fname = NULL;
 
     if (mbname->localpart) {
-        char *mboxname = mboxname_user_mbox(mbname_userid(mbname), NULL);
+        char *inboxname = mboxname_user_mbox(mbname_userid(mbname), NULL);
         mbentry_t *mbentry = NULL;
-
-        int r = mboxlist_lookup_allow_all(mboxname, &mbentry, NULL);
-        free(mboxname);
+        int r = mboxlist_lookup_allow_all(inboxname, &mbentry, NULL);
+        free(inboxname);
+
+        if (r == IMAP_MAILBOX_NONEXISTENT) {
+            // look for the INBOX of previous names in order to see if we're mid-rename
+            mbentry_t *mbentry_byname = NULL;
+            mbentry_t *mbentry_byid = NULL;
+            r = mboxlist_lookup_allow_all(mbname_intname(mbname), &mbentry_byname, NULL);
+            if (!r) r = mboxlist_lookup_by_uniqueid(mbentry_byname->uniqueid, &mbentry_byid, NULL);
+            if (!r) {
+                int i;
+                for (i = 0; i < ptrarray_size(&mbentry_byid->name_history); i++) {
+                    const former_name_t *fname = ptrarray_nth(&mbentry_byid->name_history, i);
+                    char *olduserid = mboxname_to_userid(fname->name);
+                    char *oldinboxname = mboxname_user_mbox(olduserid, NULL);
+                    r = mboxlist_lookup_allow_all(oldinboxname, &mbentry, NULL);
+                    free(oldinboxname);
+                    free(olduserid);
+                    if (r != IMAP_MAILBOX_NONEXISTENT) break;
```

Comment:
This line looks like it'll break out of the loop on success OR any other error, is that the intent?  If so it could use a comment, or maybe an explicit `if (r == 0 || r != IMAP_MAILBOX_NONEXISTENT) ...`, to make it clear to the reader that both cases have been considered and this choice is deliberate.

Or, I guess maybe mboxlist_lookup_allow_all only ever returns 0 or IMAP_MAILBOX_NONEXISTENT, so not being nonexistent means success?  Not sure, I haven't looked. Even if that is the case now, being explicit about the intent here will protect us in the future.

Reply:
I think this entire chunk of code could be eliminated if we also use #3571

Reply:
Yep, that's definitely a good plan.